lib/datadog/profiling/collectors/system_info.rb

```diff
@@ -0,0 +1,101 @@
+# frozen_string_literal: true
+
+require 'set'
+require 'json'
+
+module Datadog
+  module Profiling
+    module Collectors
+      # Collects SystemInfo data of relevance for profiler. This will get sent alongside
+      # the profile and show up in the UI or influence processing.
+      class SystemInfo
+        def initialize(settings)
+          @settings = settings
+          refresh
+        end
+
+        attr_reader :system_info
+
+        def refresh
+          @system_info = {
+            platform: collect_platform_info,
+            runtime: collect_runtime_info,
+            application: collect_application_info,
+            profiler: collect_profiler_info,
+          }
+          self
+        end
+
+        private
+
+        # Instead of trying to figure out real process start time by checking
+        # /proc or some other complex/non-portable way, approximate start time
+        # by time of requirement of this file.
+        START_TIME = Time.now.utc.freeze
+
+        def collect_platform_info
+          @platform_info ||= {
+            container_id: Datadog::Core::Environment::Container.container_id,
+            hostname: Datadog::Core::Environment::Platform.hostname,
+            kernel_name: Datadog::Core::Environment::Platform.kernel_name,
+            kernel_release: Datadog::Core::Environment::Platform.kernel_release,
+            kernel_version: Datadog::Core::Environment::Platform.kernel_version
+          }
+        end
+
+        def collect_runtime_info
+          @runtime_info ||= {
+            engine: Datadog::Core::Environment::Identity.lang_engine,
+            version: Datadog::Core::Environment::Identity.lang_version,
+            platform: Datadog::Core::Environment::Identity.lang_platform,
+          }
+        end
+
+        def collect_application_info
+          @application_info ||= {
+            start_time: START_TIME.iso8601,
+            env: @settings.env,
+            service: @settings.service,
+            version: @settings.version,
+          }
+        end
+
+        def collect_profiler_info
+          unless @profiler_info
```

Comment:
As this class reuses the info hashes it gives out on `system_info`, it may be worth sprinkling `.freeze` on all of them, just to avoid accidental mutations somewhere ;)